baseclass/src/main/java/com/google/apigee/calloutbaseclass/CalloutsBase.java

```diff
@@ -0,0 +1,196 @@
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.apigee.calloutbaseclass;
+
+import com.apigee.flow.message.MessageContext;
+
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/*
+ * Base Class for Apigee Java Callouts containing commonly used methods across callouts.
+ */
+public abstract class CalloutsBase {
+
+  private static final String ERROR_FLOW_VARIABLE = "callout_error";
+  private static final String EXCEPTION_FLOW_VARIABLE = "callout_exception";
+  private static final String EXCEPTION_STACKTRACE_FLOW_VARIABLE = "callout_exception_stacktrace";
+  private static final String LOG_FLOW_VARIABLE = "callout_log";
+
+  // Matches common error strings (e.g. matches "example: error")
+  private static final String commonErrorPatternString = "^(.+?)[:;] (.+)$";
+  private static final Pattern commonErrorPattern = Pattern.compile(commonErrorPatternString);
+  // Matches flow variable reference (e.g. matches "{request.test}")
+  private static final String variableReferencePatternString =
+      "(.*?)\\{([^\\{\\} :][^\\{\\} ]*?)\\}(.*?)";
+  private static final Pattern variableReferencePattern =
+      Pattern.compile(variableReferencePatternString);
+
+  private final Map<String, String> properties;
+
+  /**
+   * Constructor taking in String key and value properties map
+   *
+   * @param properties String key and value map
+   */
+  public CalloutsBase(Map<String, String> properties) {
+    this.properties = properties;
+  }
+
+  /**
+   * Retrieves an optional property from properties map. Returns null if not found in properties map.
+   *
+   * @param propertyName Name of property to retrieve value
+   * @param messageContext Message Context
+   * @return Value of property in properties map
+   */
+  public String getOptionalProperty(String propertyName, MessageContext messageContext) {
+    if (!this.properties.containsKey(propertyName)) {
+      return null;
+    }
+    String value = resolveVariableReferences(this.properties.get(propertyName).trim(), messageContext);
+    if ("".equals(value)) {
+      return null;
+    }
+    return value;
+  }
+
+  /**
+   * Retrieves a required property from properties map.
+   *
+   * @param propertyName Name of property to retrieve value
+   * @param messageContext Message Context
+   * @return Value of property in properties map
+   * @throws IllegalArgumentException if the propertyName does not exist in properties map or resolves to an empty string.
+   */
+  public String getRequiredProperty(String propertyName, MessageContext messageContext) {
```

Comment:
invoke getOptionalProperty and throw IllegalArgumentException when the property is not available.  you can do it another way around too but I think it's better to not deal with exceptions.  Up to you.